### ISSUE TITLE
Question about Interaction Label Calculation

### ISSUE THREAD

**Issue by biomolecule** [phdymz/ProteinMAE#7]
## Question about Interaction Label Calculation

Hello, I’m trying to reproduce the search part of your work. 
I noticed that you applied random rotation and centering to the coordinates in the dataset for data augmentation. However, instead of directly computing the interaction labels, you calculate them in the `compute_loss` function. 

When computing, you use the Euclidean distance of the rotated and centered coordinates as the label.  
However, the data augmentation in your code may disrupt the original interaction interface, meaning the calculated labels might not represent the actual surface interactions. 
Could it be that I misunderstood something?  

---

## 关于交互标签计算的问题

您好，我正在试图复现您关于 `search` 的部分。
我注意到，您在数据集中对坐标进行了随机旋转和中心化，以实现数据增强。但是，您并没有直接计算交互标签，而是在 `compute_loss` 函数中计算。  

在计算过程中，您使用旋转和中心化后的坐标的欧式距离作为标签。然而，代码中的数据增强部分可能会破坏原本的交互界面，因此据此计算出的标签可能无法准确代表真实的表面交互。
请问是我的理解出现了偏差吗？

**search dataset**
```
        xyz1 = np.matmul(rot, xyz1.T).T
        atom1 = np.matmul(rot, atom1.T).T
        normal1 = np.matmul(rot, normal1.T).T

        xyz2 = np.matmul(rot, xyz2.T).T
        atom2 = np.matmul(rot, atom2.T).T
        normal2 = np.matmul(rot, normal2.T).T

        atom_center1 = atom1.mean(0)
        xyz1 = xyz1 - atom_center1
        atom1 = atom1 - atom_center1

        atom_center2 = atom2.mean(0)
        xyz2 = xyz2 - atom_center2
        atom2 = atom2 - atom_center2
        ...
        return xyz1, normal1, label1, curvature1, dists1, atom_type_sel1, xyz2, normal2, label2, curvature2, dists2, atom_type_sel2
```

**iterate**
```
    for it, (xyz1, normal1, label1, curvature1, dist1, atom_type1, xyz2, normal2, label2, curvature2, dist2, atom_type2) in enumerate(
        tqdm(dataset)
    ):  # , desc="Test " if test else "Train")):
        ...
        outputs1 = net(xyz1, normal1, curvature1, dist1, atom_type1)
        outputs2 = net(xyz2, normal2, curvature2, dist2, atom_type2)

        loss, sampled_preds, sampled_labels  = compute_loss(outputs1, outputs2, xyz1, xyz2, label1, label2)
```

**compute_loss**
```
def compute_loss(descs1, descs2, xyz1, xyz2, label1, label2):

    loss = 0
    preds_concates = []
    labels_concates = []

    for i in range(len(label1)):
        idx1 = label1[i] > 0
        idx2 = label2[i] > 0

        pos_descs1 = descs1[i][idx1]
        pos_descs2 = descs2[i][idx2]

        pos_xyz1 = xyz1[i][idx1]
        pos_xyz2 = xyz2[i][idx2]

        pos_xyz_dists = (
            ((pos_xyz1[:, None, :] - pos_xyz2[None, :, :]) ** 2).sum(-1).sqrt()
        )
        pos_desc_dists = torch.matmul(pos_descs1, pos_descs2.T)

        pos_preds = pos_desc_dists[pos_xyz_dists < 1.0]
        pos_labels = torch.ones_like(pos_preds)
        ...
```

**Comment by phdymz** [comment]
we follow the settings in dmasif

**Comment by biomolecule** [comment]
## Question about Point Cloud Transformation and Restoration in ProteinMAE

Thank you for your answer. I found that the point cloud transformation in dMasif includes random rotation and centering. 
However, during iteration, if it's a search task, dMasif will use the saved rotation matrix and center information in the `generate_matchinglabels` function to restore the xyz coordinates and then calculate the label and loss.  

I noticed that in ProteinMAE, you placed the corresponding transformation steps in the dataset. How do you handle this restoration operation?

---

## 关于 ProteinMAE 中点云变换和恢复操作的问题

感谢你的解答。我发现，dMasif 的点云变换包括随机旋转和中心化。
然而，在迭代过程中，如果是 search 任务，dMasif 会在函数`generate_matchinglabels`中使用保存的旋转矩阵和中心信息来恢复 xyz 坐标，然后计算标签和损失。  

我注意到，在 ProteinMAE 中，您将对应的变换步骤放在了 dataset 部分。请问，您是如何处理这个恢复操作的呢？

**dmasif/data_iteration.py**
```
def iterate(
    net,
    dataset,
    optimizer,
    args,
    test=False,
    save_path=None,
    pdb_ids=None,
    summary_writer=None,
    epoch_number=None,
):
    """Goes through one epoch of the dataset, returns information for Tensorboard."""

            ...

            if args.random_rotation:
                P1["rand_rot"] = protein_pair.rand_rot1.view(-1, 3, 3)[0]
                P1["atom_center"] = protein_pair.atom_center1.view(-1, 1, 3)[0]
                P1["xyz"] = P1["xyz"] - P1["atom_center"]
                P1["xyz"] = (
                    torch.matmul(P1["rand_rot"], P1["xyz"].T).T
                ).contiguous()
                P1["normals"] = (
                    torch.matmul(P1["rand_rot"], P1["normals"].T).T
                ).contiguous()
                if not args.single_protein:

                    ...        
                    
            torch.cuda.synchronize()
            prediction_time = time.time()
            outputs = net(P1, P2)
            torch.cuda.synchronize()
            prediction_time = time.time() - prediction_time

            P1 = outputs["P1"]
            P2 = outputs["P2"]

            if args.search:
                generate_matchinglabels(args, P1, P2)

            if P1["labels"] is not None:
                loss, sampled_preds, sampled_labels = compute_loss(args, P1, P2)
            else:
                loss = torch.tensor(0.0)
                sampled_preds = None
                sampled_labels = None
```

**generate_matchinglabels**
```
def generate_matchinglabels(args, P1, P2):
    if args.random_rotation:
        P1["xyz"] = torch.matmul(P1["rand_rot"].T, P1["xyz"].T).T + P1["atom_center"]
        P2["xyz"] = torch.matmul(P2["rand_rot"].T, P2["xyz"].T).T + P2["atom_center"]
    xyz1_i = LazyTensor(P1["xyz"][:, None, :].contiguous())
    xyz2_j = LazyTensor(P2["xyz"][None, :, :].contiguous())

    xyz_dists = ((xyz1_i - xyz2_j) ** 2).sum(-1).sqrt()
    xyz_dists = (1.0 - xyz_dists).step()

    p1_iface_labels = (xyz_dists.sum(1) > 1.0).float().view(-1)
    p2_iface_labels = (xyz_dists.sum(0) > 1.0).float().view(-1)

    P1["labels"] = p1_iface_labels
    P2["labels"] = p2_iface_labels
```
